designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+
+
+### Q: How should Karpenter handle the default or undefined reason case? 
+If a budget reason is unspecified like budgets[1], we will assume this budget applies to all actions that are not specified 
+```yaml
+budgets: 
+  - nodes: 10
+    reasons: [drifted, underutilized]
+    schedule: "* * * * *"
+  - nodes: 30 
+    schedule: "* * * * *" 
+```
+Meaning that for any actions other than drifted + consolidation, the total amount of disrupting + unhealthy nodes has to be less than 30 for them to trigger disruption.
+
+### Q: How should allowed disruptions be calculated in a multi-reason world? 
+The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:
+
+AllowedDisruptions = minNodeCountOfAnActiveBudget - unhealthyNodes - totalDisruptingNodes.
+
+When calculating by reason, two potential equations emerge:
+1. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes[reason]
+2. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes
+The second equation is the one we should opt into.
+
+Take this budget as an example
+```yaml
+budgets: 
+  - nodes: 15 
+    reasons: [drifted, underutilized]
+    schedule: "* * * * *"
+  - nodes: 10
+    reasons: [drifted]
+    schedule: "* * * * *"
+  - nodes: 5 
+    schedule: "* * * * *" 
+```
+First, calculate minNodeCountOfActiveBudget. Given there are two budgets for drifted, we select the lower number:
+
+```
+minNodeCountOfActiveBudget = {
+  drifted: 10 
+  underutilized: 15 
+  Default: 5
+}
+```
+
+In the first equation, tracking the number of nodes disrupted by each reason is required. For example:
+```
+disrupting = {
+  drifted: 3 
+  underutilized: 6
+  Expriation: 3
+  empty: 2
+}
+```
+
+Assuming zero unhealthy nodes:
+**First Equation**
+- drifted = 10 - 3: Disruption allowed as drifted bucket isn't full.
+- underutilized = 15 - 6: Disruption allowed since 15 - 6 > 0.
+- Default = 5 - expired - empty = 0: No disruption allowed for reasons other than consolidation and drift.
+
+**Second Equation**
+- Total Disrupting = 3 + 6 + 5 = 14.
+- drifted = 10 - 14: No disruption allowed due to drift.
+- underutilized = 15 - 14: Disruption allowed for at least one node due to consolidation.
+- Default: 5 - 14: No disruption allowed for other methods.
+
+#### Decision
+The second equation simplifies reasoning for cluster operators. A node can be tagged for multiple disruption reasons (e.g., empty, drifted, underutilized), making it challenging to precisely calculate in-flight disruptions. The second equation streamlines this process by providing a clearer view of the overall disruption impact.
+### Q: How should we handle an unspecified default reason? 
```

Comment:
```suggestion
### Q: How should we handle an unspecified reason when others are specified? 
```

I was a little unclear on this when I first read it. Maybe tightening up the language of the question could help here. From what I understand, you are asking how do we handle the case for expiry below since it's not specified in the budget.